internal/state/configuration.go

```diff
@@ -72,45 +85,47 @@ func buildConfiguration(graph *graph) Configuration {
 		return Configuration{}
 	}
 
-	configBuilder := newConfigBuilder()
+	upstreams := buildUpstreams(graph.Backends)
+
+	builderFactory := newVirtualServerBuilderFactory()
 
 	for _, l := range graph.Gateway.Listeners {
 		// only upsert listeners that are valid
 		if l.Valid {
-			configBuilder.upsertListener(l)
+			builder := builderFactory.getBuilderForProtocol(l.Source.Protocol)
+			builder.upsertListener(l)
 		}
 	}
 
-	return configBuilder.build()
+	config := Configuration{
+		HTTPServers: builderFactory.getBuilderForProtocol(v1beta1.HTTPProtocolType).build(),
+		SSLServers:  builderFactory.getBuilderForProtocol(v1beta1.HTTPSProtocolType).build(),
+		Upstreams:   upstreams,
+	}
+
+	return config
 }
 
-type configBuilder struct {
```

Comment:
This doesn't appear to be a Builder or a Factory.

To be a Builder I'd expect a concrete object and an abstract builder.
```
type Builder
type Concrete

func NewBuilder() Builder // returns Builder
func (Builder) AddX() Builder // keeps returning itself as a Builder
func (Builder) AddY() Builder // keeps returning itself as a Builder
func (Builder) Build() Concrete // final stage gives us the composite object
```

Or a Factory, where one abstract method can give us multiple implementations.

I'm concerned these names diverge from their accepted patterns.

Reply:
I think it's getBuilderForProtocol is confusing. Shouldn't the Factory return the correct Builder by accepting arguments.

Reply:
Valid points. I was using the name `builder` because of the intent of the package, which is to build configuration. However, I can see how that can be misleading. 

As far as the factory goes, after another look, I don't think the pattern fits the use case -- it just overcomplicates things.

I've replaced the factory and builder with what I think is a more straightforward approach. Let me know what you think.